cpp/src/arrow/dataset/dataset.h

```diff
@@ -59,6 +156,17 @@ class ARROW_DS_EXPORT Fragment : public std::enable_shared_from_this<Fragment> {
   virtual Result<RecordBatchGenerator> ScanBatchesAsync(
       const std::shared_ptr<ScanOptions>& options) = 0;
 
+  /// \brief Inspect a fragment to learn basic information
```

Comment:
This is basically to read the footer ahead of time? 

So, ParquetFileFragment would attach the actual Parquet footer in the structure it returns, the Arrow fragment would return the schema, the CSV fragment would throw its hands up, etc.?

Reply:
Yes, exactly.  I'm not a huge fan of this because now fragment scanning is broken into three steps...

FragmentScanner: Inspect fragment (new step)
Scanner: Create evolution
FragmentScanner: Open reader (where we previously determined column names / schema)
Scanner: Start iteration task
FragmentScanner: Scan

I could push the evolution creation into the fragment scanner but that makes a new step that every fragment scanner has to do and so I think that would be worse.

Your point on "actual parquet footer" and "the schema" is also correct.  My "prototypical future evolution strategy" that I am trying to make sure we plan for is to resolve field references by parquet column id.  So it would work in this fashion:

 * User specifies numeric field references against the dataset schema
 * Evolution strategy has a lookup table from numeric position in dataset schema  to column id
 * On inspection, physical column ids are inserted into the schema and passed to the evolution
 * Evolution uses the lookup table combined with the physical column ids to determine the physical columns that are being asked for using
 
 So this way the inspect step would need to return something with the column IDs.  This could be the parquet footer.  However, it also could be the schema extracted from the footer as we have a mechanism for encoding those column IDs into the schema today.  This way we could leave the door open to someone adding column IDs to IPC files as well.

Reply:
Sounds good to me.

The Parquet/Arrow readers, IIRC, let you open a file and provide an already-parsed footer to at least skip that step.

Reply:
I don't see the problem with splitting scanning into these steps, FWIW. I think this would also let us down the line support schemes where metadata (footers) are cached in an index separately from the actual data (which would let us skip a synchronous I/O step)